### DISCUSSION TITLE
Modbus read/write HOLDING_REGISTERS or COILS.

### DISCUSSION THREAD

**Discussion by levin8667** [edgexfoundry/community#160] [Q&A] [answered]
Hi all. I am conducting modbus device testing and hope to connect to edgex. There are a few questions that have confused me. If anyone knows how to do this, I will be grateful I look forward to your valuable comments, have a great day.

1. Can device-modbus read single coil? I did a test and found that the rawType is set to int16. The device must open at least 16 coils to successfully read. However, out of the data obtained from a read command, 15 coils are useless to me. I want to know why the rawType only supports int16/uint16. What confuses me is that my device only has coils 0 to 3, and I cannot access the device using the device-modbus service. When I sent a GET request, I received an incorrect response: "request failed, status code: 500, err: {\"apiVersion\":\"v3\",\"message\":\"error reading Regex DeviceResource(s) Temperature for test-modbus -\\u003e modbus: exception '2' (illegal data address), function '129'\",\"statusCode\":500}".

2. Can device-modbus write a single coil? When I sent a SET request, I received an incorrect response: "code: 503, message: failed to send a HTTP request".

3. Can device-modbus read a certain bit of HOLDING_REGISTERS. A register has 16 bits, and I have a collection device whose 16 switch values are stored in 16 bits of a register. Can I configure this information in the Device Profile yml file?

4. Similarly, can device-modbus write a certain bit of HOLDING_REGISTERS?  because I need to control 16 switch values.

**Comment by cloudxxx8** [comment, marked as answer]
@levin8667 
1 and 2 are easy.  You should not define rawType for Coils.  The design of rawType is for integer and float data type conversion with scales.
https://docs.edgexfoundry.org/3.0/examples/Ch-ExamplesModbusdatatypeconversion/
For the normal usage of Coils, please just define valueType to Bool without defining the rawType, and it will work as your expectation.

For 3 and 4, the community version of device-modbus doesn't support this feature, but it's doable.  You can open an issue and write down your feature request and see if there is any volunteer to help.  If you know how to do and want to contribute to the community, we are welcome you to just open a PR.
Another option is to adopt the commercial version of device-modbus.
The commercial device-modbus provided by IOTech support this feature.  Please see `boolIndex` attribute.
https://docs.iotechsys.com/edge-xpert23/device-services/modbus/modbus-overview.html#optional-attributes

**Reply by levin8667** [reply]
Thank you for your reply!
